### PR TITLE
Segment Path Filter & Clickfunnels Button

### PR DESCRIPTION
**Pull request by ripxip**
Added a path filter plugin that parses URL paths from learn.xip.co and adds relevant properties to the event object. It checks the URL path for a specific pattern and extracts the relevant information to enrich the event object.

Also added a event listener that listens for a click event on a .elButton element and tracks that as a button click with Segment. aadditional browsing data and the text within the button are added to the event as context.

### CODE REVIEW

**Comment by ripxip** [comment]
@eddiesiegel I believe this is ready for review

**Comment by ripxip** [comment]
@ravi-xip Do you mind reviewing this? Eddie is pinched for time and I figured I could get a first pass on reviewing from you.

File: src/clickfunnels/main.ts
Comment (ravi-xip) [inline src/clickfunnels/main.ts]: Remove console.log

File: src/common/segment-utils.ts
Comment (ravi-xip) [inline src/common/segment-utils.ts]: What does the type: 'enrichment' mean here?

File: src/common/segment-utils.ts
Comment (ravi-xip) [inline src/common/segment-utils.ts]: Would this be the only case? What does -page in the path refer to here?

File: src/common/segment-utils.ts
Comment (ravi-xip) [inline src/common/segment-utils.ts]: A simpler way to write this could be:
```js
if not (isBetween(pathComponents.length, 3, 4)) return ctx;
....
```

File: src/clickfunnels/main.ts
Comment (ravi-xip) [inline src/clickfunnels/main.ts]: The term `additionalData` seems a bit vague to me.

**Review by ravi-xip — APPROVED** [review]
Overall, LGTM!

File: src/common/segment-utils.ts
Comment (eddiesiegel) [inline src/common/segment-utils.ts]: Why are we calling this urlPath**Filter**? It's not filtering anything, right? What about `ClickfunnelsUrlEnrichmentPlugin` or something like that?

Also: I think it's better to move this somewhere else. You have put it in `common/segment-utils.ts` but it's a segment plugin that's very specific to clickfunnels (doesn't apply to webflow, e.g.). I would suggest putting it either (a) in a clickfunnels util file like `clickfunnels/segment-utils.ts` or (b) in a file that is all its own w/the same name as the class (like if you end up calling it `ClickfunnelsUrlEnrichmentPlugin` then you could put it in `clickfunnels/url-enrichment-plugin.ts` or similar)

File: src/common/segment-utils.ts
Comment (eddiesiegel) [inline src/common/segment-utils.ts]: Why are we using `ctx.event.properties.url` and not `window.location`?

File: src/common/segment-utils.ts
Comment (eddiesiegel) [inline src/common/segment-utils.ts]: Is there a reason why you added this check? e.g., they provide us with a test site, right? Don't we want the enrichment to work there too? Any reason to restrict the URL at all?

File: src/common/segment-utils.ts
Comment (eddiesiegel) [inline src/common/segment-utils.ts]: Can you refine this comment? This comment really just says exactly what the next line says without telling us anything new / not obvious. I genuinely don't know what this `-page` convention is going to mean, so I think a comment could be very helpful. Something like "we end certain paths with '-page' to indicate blah blah, for example [example-path], so we handle that here by trimming off any trailing '-page'"

File: src/common/segment-utils.ts
Comment (eddiesiegel) [inline src/common/segment-utils.ts]: Same here. When are we going to have a `-top-level` page, what does it mean, and what are we doing about it here?

Example to get you started (feel free to discard):

"If the path ends with -top-level then it means xyz, so we trim off the '-top-level' from the end and treat the beginning as the name of the funnel. For these pages, we just append a single "funnel" property to our event and return early without attempting to add all of the other properties."

File: src/common/segment-utils.ts
Comment (eddiesiegel) [inline src/common/segment-utils.ts]: Can you add more comments here explaining exactly what properties we're going to be adding to each Segment track call?

File: src/common/segment-utils.ts
Comment (eddiesiegel) [inline src/common/segment-utils.ts]: I feel like we could use comments at several places in this function explaining what the path could look like. There's all this specific string manipulation going on and I have no idea what the string is expected to look like.

File: src/common/segment-utils.ts
Comment (eddiesiegel) [inline src/common/segment-utils.ts]: What is this `.replace` meant to handle? Just a leading `/` at the front of the path? Maybe it should have been done earlier then - strip the slash before parsing it into its components so that this later code doesn't have to be aware of the fact that `instructor` came first.

File: src/common/segment-utils.ts
Comment (eddiesiegel) [inline src/common/segment-utils.ts]: Nit: Not important, but FYI - newer versions of javascript have nicer syntax for setting values on an object when the name of the variable is the same as the key. The following two things are equivalent:

```javascript
const myObj = {
  x: x,
  y: y,
  z: z
}
```
and
```javascript
const myObj = {
  x,
  y,
  z
}
```

So in your example, if you rename the `version` variable to be `funnelVersion` and implement my other suggestion to strip the leading `/` off of the path, then you'll be able to do this:

```javascript
ctx.event.properties = {
  ...ctx.event.properties, instructor, funnelType, funnelStep, funnelVersion
}
```

File: src/clickfunnels/main.ts
Comment (eddiesiegel) [inline src/clickfunnels/main.ts]: Two problems with this:
1. `querySelector` returns only one element. So we'll only be listening to clicks on a single button (and if there are multiple buttons on the page it'll pick a random one). You're most likely looking for `querySelectorAll`
2. This will only work for buttons that are already on the page at the time the page loads. Is it possible for clickfunnels to dynamically add stuff to the page, or is everything always loaded on the page and dynamically shown/hidden? If you're concerned about elements being added dynamically then you need to register a `MutationObserver` that registers new click handlers every time a new button is added.

File: src/clickfunnels/main.ts
Comment (eddiesiegel) [inline src/clickfunnels/main.ts]: Is this something we should be attaching to every event? Maybe we should just register a Plugin w/segment to add these attrs to everything on all of our pages?

File: src/clickfunnels/main.ts
Comment (eddiesiegel) [inline src/clickfunnels/main.ts]: Nit: Similar feedback to my other comment. You can do `{buttonText, queryParams, ...additionalData}`

Comment (eddiesiegel) [reply src/clickfunnels/main.ts]: EDIT: Actually why don't we just add an ESLint rule for this? Can you try adding this to our `.eslintrc` and seeing if it helps us be more consistent w/this?

```json
{
  "rules": {
    "object-shorthand": ["error", "properties"]
  }
}

```

File: src/clickfunnels/main.ts
Comment (eddiesiegel) [inline src/clickfunnels/main.ts]: I think this needs to be called before `enableSegment` gets called so that we can make sure the right things are appended to the `.page()` call and anything else that `enableSegment` does automatically internally.

Maybe we could avoid this trouble in the future by instead updating `enableSegment` to take an optional list of plugins as arguments and letting `enableSegment` take care of the correct order of operations internally.

So here in `main.ts` you could just do `enableSegment('YTsllNl2tO9CSJ8OqG7qtz2EW00HElZG', urlPathFilter)` and it'll just work.

**Review by eddiesiegel — CHANGES_REQUESTED** [review]

File: src/common/segment-utils.ts
Comment (ripxip) [reply src/common/segment-utils.ts]: https://github.com/segmentio/analytics-next/tree/master/packages/browser#-plugins
It's mentioned here as a type for the plugin in segment. It's used in the case that you want to 'enrich' event data using a plugin.

File: src/common/segment-utils.ts
Comment (ripxip) [reply src/common/segment-utils.ts]: Because I wanted to use the URL from the event and not from the site. Thinking about it now, I'm not sure it matters. Is there a reason we should go either way? Would the two values ever not be equivalent?

File: src/common/segment-utils.ts
Comment (ripxip) [reply src/common/segment-utils.ts]: updated with a little explanation

File: src/common/segment-utils.ts
Comment (ripxip) [reply src/common/segment-utils.ts]: updated with an explanation. Clickfunnels has a page and a funnel page in short. I wanted to make sure we handle both cases.

File: src/common/segment-utils.ts
Comment (ripxip) [reply src/common/segment-utils.ts]: added some more detailed comments

File: src/common/segment-utils.ts
Comment (ripxip) [reply src/common/segment-utils.ts]: Added some url examples to the comments

File: src/common/segment-utils.ts
Comment (ripxip) [reply src/common/segment-utils.ts]: added this to the setting of the pathComponents variable instead.

File: src/common/segment-utils.ts
Comment (ripxip) [reply src/common/segment-utils.ts]: gotcha, I implemented it that way. Thanks for the heads up

File: src/common/segment-utils.ts
Comment (ripxip) [reply src/common/segment-utils.ts]: I don't think inBetween is a built in function. I'm leaving this as is since it's too simple for me to break out into it's own function.

File: src/common/segment-utils.ts
Comment (ripxip) [reply src/common/segment-utils.ts]: I agree. I moved this to clickfunnels/segment-utils.ts

File: src/common/segment-utils.ts
Comment (ripxip) [reply src/common/segment-utils.ts]: Going to do that in clickfunnels/segment-utils.ts. Check there for a more descriptive comment.

File: src/clickfunnels/main.ts
Comment (ripxip) [reply src/clickfunnels/main.ts]: Added this functionality to enableSegment function. Please check on that to see if I did it right. Copilot seems to think that I am doing it correctly so I feel good about it.

File: src/clickfunnels/main.ts
Comment (ripxip) [reply src/clickfunnels/main.ts]: Not going to change this because I think I'll be breaking this out into a plugin, thus removing the need for this variable name

File: src/clickfunnels/main.ts
Comment (ripxip) [reply src/clickfunnels/main.ts]: Added this to .eslintrc.js

File: src/clickfunnels/main.ts
Comment (ripxip) [reply src/clickfunnels/main.ts]: The clickfunnels enrichment plugin relies on this url and path data. Do you know if there's a way to guarantee that this as a plugin would run before the clickfunnels enrichment plugin?

Comment (ripxip) [reply src/clickfunnels/main.ts]: Made this a function in common/segment-utils.ts instead of a plugin

File: src/clickfunnels/main.ts
Comment (ripxip) [reply src/clickfunnels/main.ts]: Broke the listener attachment out into a function and use querySelectorAll as well as a mutation observer

File: src/common/segment-utils.ts
Comment (ripxip) [reply src/common/segment-utils.ts]: Changed to window.location.href. I think this would work.

Comment (ripxip) [reply src/common/segment-utils.ts]: jk I was looking at the wrong spot. Keeping this as ctx.event.properties since it's within the context of the plugin, which I want to only deal with event data

File: src/common/segment-utils.ts
Comment (ripxip) [reply src/common/segment-utils.ts]: Since I added the plugin to clickfunnels/segment-utils.js I don't think there's any reason to restrict anymore. This should only be called when a clickfunnels url exists anyways.

File: src/clickfunnels/main.ts
Comment (ravi-xip) [inline src/clickfunnels/main.ts:89]: nit: Can move learn.xip to a constant to avoid repititon.

File: src/clickfunnels/main.ts
Comment (ravi-xip) [inline src/clickfunnels/main.ts]: Is there a case where the querySelectorAll can return null?

File: src/clickfunnels/main.ts
Comment (ravi-xip) [inline src/clickfunnels/main.ts]: nit: Can rename to buttonMutationObserver.

**Review by ravi-xip — APPROVED** [review]
LGTM!

File: src/clickfunnels/main.ts
Comment (eddiesiegel) [inline src/clickfunnels/main.ts]: Not important at all, just an FYI: since the expected argument to `forEach` has the same method signature as `addButtonClickListener`, you can omit the anonymous function that you're wrapping around it.

```suggestion
  document.querySelectorAll('.elButton').forEach(addButtonClickListener);
```

File: src/clickfunnels/main.ts
Comment (eddiesiegel) [inline src/clickfunnels/main.ts]: I've run into this issue in the past and I think it's possible that this won't catch every new `.elButton` on the page.

I'm a little rusty on this, but I think that `mutation.addedNodes` will tell you any top-level nodes that were just added to the page, but it's possible that those nodes have children. So if a node that is not an `.elButton` gets added to the page (e.g., a form) and it has an `.elButton` as a descendent then I think your code will miss that.

You should ask chatgpt about it to see if my understanding is correct. If I'm right, then this edge case keeps coming up, so it's probably worth writing a helper function for this. something that we can call like `registerLiveEventHandler('.elButton', 'click', myCallback)`that will internally automatically handle registering our callback with every element that matches the selector both now and in the future.

FYI back in the days of jquery, this was one of the more useful functions that it provided: https://api.jquery.com/live/

File: src/clickfunnels/main.ts
Comment (eddiesiegel) [reply src/clickfunnels/main.ts]: Looks good to me :-)

File: src/clickfunnels/segment-utils.ts
Comment (eddiesiegel) [inline src/clickfunnels/segment-utils.ts]: Nit: you don't really need this else. The very next thing that happens is you `return ctx` anyway

**Review by eddiesiegel — CHANGES_REQUESTED** [review]
Overall this looks good! There is one bug that needs to be addressed before this is ready to ship (and a few nits that you can either address or ignore). Once that bug is addressed you can feel free to ship without another approval if you prefer not to wait.

File: src/clickfunnels/main.ts
Comment (ripxip) [reply src/clickfunnels/main.ts]: Added a registerLiveEventHandler in common/segment-utils.js. I can probably move that out of there is necessary. In this file I'm only defining the event handler and passing that into registerLiveEventHandler.

File: src/clickfunnels/main.ts
Comment (ripxip) [reply src/clickfunnels/main.ts]: not only an empty node list.

**Comment by ripxip** [comment]
Merging as all comments are resolved from second review pass.

**Pull request merged**
